sympy/sets/sets.py

```diff
@@ -2027,6 +2027,168 @@ def __iter__(self):
+    DisjointUnion(FiniteSet(1, 2, 3), Interval(0, 5))
+    >>> DisjointUnion(A, B).rewrite(Union)
+    Union(ProductSet(FiniteSet(1, 2, 3), FiniteSet(0)), ProductSet(Interval(0, 5), FiniteSet(1)))
+    >>> C = FiniteSet(Symbol('x'), Symbol('y'), Symbol('z'))
+    >>> DisjointUnion(C, C)
+    DisjointUnion(FiniteSet(x, y, z), FiniteSet(x, y, z))
+    >>> DisjointUnion(C, C).rewrite(Union)
+    ProductSet(FiniteSet(x, y, z), FiniteSet(0, 1))
+
+    References
+    ==========
+
+    https://en.wikipedia.org/wiki/Disjoint_union
+    """
+
+    def __new__(cls, *sets):
+        dj_collection = []
+        for set_i in sets:
+            if isinstance(set_i, Set):
+                dj_collection.append(set_i)
+            else:
+                raise ValueError("Invalid input: '%s', input args \
+                    to DisjointUnion must be Sets" % set_i)
+        obj = Basic.__new__(cls, *dj_collection)
+        return obj
+
+    @property
+    def sets(self):
+        return self.args
+
+    @property
+    def is_empty(self):
+        return fuzzy_and(s.is_empty for s in self.sets)
+
+    @property
+    def is_finite_set(self):
+        all_finite = fuzzy_and(s.is_finite_set for s in self.sets)
+        return fuzzy_or([self.is_empty, all_finite])
+
+    @property
+    def is_iterable(self):
+        if self.is_empty:
+            return False
+        iter_flag = True
+        for set_i in self.sets:
+            if not set_i.is_empty:
+                iter_flag = iter_flag and set_i.is_iterable
+        return iter_flag
+
+    def _eval_rewrite_as_Union(self, *sets):
+        """
+        Rewrites the disjoint union as the union of (``set`` x {``i``})
+        where ``set`` is the element in ``sets`` at index = ``i``
+        """
+
+        dj_union = EmptySet()
+        index = 0
+        for set_i in sets:
+            if isinstance(set_i, Set):
+                cross = ProductSet(set_i, FiniteSet(index))
+                dj_union = Union(dj_union, cross)
+                index = index + 1
+            else:
+                continue
+        return dj_union
+
+    def _contains(self, element):
+        """
+        'in' operator for DisjointUnion
+
+        Examples
+        ========
+
+        >>> from sympy import Interval, DisjointUnion
+        >>> D = DisjointUnion(Interval(0, 1), Interval(0, 2))
+        >>> (0.5, 0) in D
+        True
+        >>> (0.5, 1) in D
+        True
+        >>> (1.5, 0) in D
+        False
+        >>> (1.5, 1) in D
+        True
+
+        Passes operation on to constituent sets
+        """
+        if not isinstance(element, Tuple) or len(element) != 2:
+            return False
+
+        if not element[1].is_Integer:
+            return False
+
+        if element[1] >= len(self.sets) or element[1] < 0:
+            return False
+
+        return element[0] in self.sets[element[1]]
+
+    def __iter__(self):
+        if self.is_iterable:
+            return iter(self.rewrite(Union))
+        #     index = 0
```

Comment:
Why did you comment out this? I think that it should be removed in a production code if it is not going to be implemented

Reply:
Ah, yes. I request your input on this. Even though I have implemented `DisjointUnion` as a different symbolic object, it is equivalent to `DisjointUnion.rewrite(Union)`, so should I write an `__iter__` method for `DisjointUnion` from scratch, or is the current implementation fine, i.e., using the iterator of `self.rewrite(Union)`?
Strictly speaking, the current implementation of `__iter__` will return elements of `self.rewrite(Union)`, which are by choice of implementation, the same as the elements of `self`.

Reply:
There would better need a guard against `self.rewrite(Union)`
because that API is not always intended to succeed rewriting to an another object or not
```
new = self.rewrite(Union)
if isinstance(new, Union):
   ...
```

Reply:
I'm sorry Sir, but I don't understand? Could you please elaborate a little more?

Reply:
But I think that it's fine defining more simple iterator under each set objects, as long as it is tested well.

Reply:
The point I mean about is that sympy rewrite is an API that is not guaranteed to succeed 100% times
e.g. it can accept nonsensical rewrite like `sin(x).rewrite(gamma)` and can return identical object

Reply:
Okay Sir, I do understand your point in the general scenario. But that should happen only if the corresponding rewrite is not defined, i.e., `sin(x).rewrite(gamma)` returns `sin(x)` because the `_eval_rewrite_as_gamma` would probably not have been defined for `sin(x)`.
In my case however, self is always a pointer to an object of `DisjointUnion`, and for that, `_eval_rewrite_as_Union` is always defined, so `self.rewrite(Union)` will always be an object of the class `Union` right?
I do however agree with you, that the direct implementation of iterator (in terms of the constituent sets) would probably be better.

Reply:
Done. 
[https://github.com/sava-1729/sympy/blob/6beae17ab87e8b4628cce120138d8b40276fcd4c/sympy/sets/sets.py#L2140-L2166](url)
Please review.